cpp/src/io/parquet/page_enc.cu

```diff
@@ -1479,6 +1524,172 @@ __global__ void __launch_bounds__(1024)
+                                         const statistics_val& v2)
+{
+  switch (ptype) {
+    case Type::BOOLEAN: return compare(v1.u_val, v2.u_val);
+    case Type::INT32:
+    case Type::INT64:
+      switch (ctype) {
+        case ConvertedType::UINT_8:
+        case ConvertedType::UINT_16:
+        case ConvertedType::UINT_32:
+        case ConvertedType::UINT_64: return compare(v1.u_val, v2.u_val);
+        default:  // assume everything else is signed
+          return compare(v1.i_val, v2.i_val);
+      }
+    case Type::FLOAT:
+    case Type::DOUBLE: return compare(v1.fp_val, v2.fp_val);
+    case Type::BYTE_ARRAY: return static_cast<string_view>(v1.str_val).compare(v2.str_val);
+    case Type::FIXED_LEN_BYTE_ARRAY:
+      if (ctype == ConvertedType::DECIMAL) return compare(v1.d128_val, v2.d128_val);
+  }
+  // calling is_comparable() should prevent reaching here
+  CUDF_UNREACHABLE("Trying to compare non-comparable type");
+  return 0;
+}
+
+/**
+ * @brief Determine if a set of statstistics are in ascending order.
+ */
+static __device__ bool is_ascending(const statistics_chunk* s,
+                                    int8_t ptype,
+                                    int8_t ctype,
+                                    uint32_t num_pages)
+{
+  for (uint32_t i = 1; i < num_pages; i++) {
+    if (compare_values(ptype, ctype, s[i - 1].min_value, s[i].min_value) > 0 ||
+        compare_values(ptype, ctype, s[i - 1].max_value, s[i].max_value) > 0)
+      return false;
+  }
+  return true;
+}
+
+/**
+ * @brief Determine if a set of statstistics are in descending order.
+ */
+static __device__ bool is_descending(const statistics_chunk* s,
+                                     int8_t ptype,
+                                     int8_t ctype,
+                                     uint32_t num_pages)
+{
+  for (uint32_t i = 1; i < num_pages; i++) {
+    if (compare_values(ptype, ctype, s[i - 1].min_value, s[i].min_value) < 0 ||
+        compare_values(ptype, ctype, s[i - 1].max_value, s[i].max_value) < 0)
+      return false;
+  }
+  return true;
+}
+
+/**
+ * @brief Determine the ordering of a set of statistics.
+ */
+static __device__ int32_t calculate_boundary_order(const statistics_chunk* s,
+                                                   int8_t ptype,
+                                                   int8_t ctype,
+                                                   uint32_t num_pages)
+{
+  if (not is_comparable(ptype, ctype)) return BoundaryOrder::UNORDERED;
+  if (is_ascending(s, ptype, ctype, num_pages))
+    return BoundaryOrder::ASCENDING;
+  else if (is_descending(s, ptype, ctype, num_pages))
+    return BoundaryOrder::DESCENDING;
+  else
+    return BoundaryOrder::UNORDERED;
+}
+
+// blockDim(1, 1, 1)
+__global__ void __launch_bounds__(1)
+  gpuEncodeColumnIndexes(device_span<EncColumnChunk> chunks,
+                         device_span<statistics_chunk const> column_stats)
```

Comment:
So this kernel gets invoked as a single thread but it seems to be doing per-page work.  Is the number of pages ever large enough that might make sense to parallelize this work a bit?  Like maybe:

- 1024 threads in 1 block, 1 thread per page storing result in shared, with t0 writing things out at the end
- A 2-kernel arrangement that goes wide on the pages, stores intermediates in global and then a second pass to do the writing?

It doesn't look like there's _too_ much work being done per page, but maybe if we have a zillion pages there's some wins.

Reply:
this is a bit like gpuEncodePageHeaders, where t0 does all the work and the other threads just wait for the sync up at the end.  In the profiling I've done, the cost of this step is very small (1ms vs 37 ms for encode pages and 400ms for snap_kernel).  I don't know if the juice would be worth the squeeze.  Can definitely revisit later if need be.